cloudify_aws/ec2/resources/routetable.py

```diff
@@ -185,3 +188,31 @@ def detach(ctx, iface, resource_config, **_):
         for association_id in association_ids:
             params.update({ASSOCIATION_ID: association_id})
             iface.detach(params)
+
+
+def attach_as_main(ctx, **_):
+    if ctx.source.node.properties['use_external_resource']:
+        targets = utils.find_rels_by_node_type(ctx.source.instance, VPC_TYPE)
+        if not targets:
+            raise NonRecoverableError(
+                "Asked to attach routing table to VPC as a main route table, "
+                "however no relationship exists to a VPC node template")
+        # len(targets) > 1 is impossible as this relationship is contained_in.
+        vpc_id = targets[0].target.instance.runtime_properties[
+            EXTERNAL_RESOURCE_ID]
+        client = Boto3Connection(ctx.source.node).client('ec2')
+        route_tables = client.describe_route_tables(
+            Filters=[{
+                "Name": "vpc-id",
+                "Values": [vpc_id]
+            }])['RouteTables']
+        main_route_table_id = None
+        for route_table in route_tables:
+            for association in route_table.get('Associations', []):
+                if association.get('Main'):
+                    main_route_table_id = route_table['RouteTableId']
+        if not main_route_table_id:
+            raise NonRecoverableError(
+                "VPC '%s' does not have a main routing table" % vpc_id)
+        ctx.source.instance.runtime_properties[EXTERNAL_RESOURCE_ID] = \
+            main_route_table_id
```

Comment:
@isaac-s with this change you are overriding the `EXTERNAL_RESOURCE_ID` for the current node template for the routing table node. Basically you are using the node routing able that is contained in a VPC to check what are the routing tables associated with that VPC so that you can get them all and get the one marked as main so that you can get its id. do not you think this is a little bit dangerous if anyone tries to reference the original routing table id then it will be something else since you override it here ?

Cannot you do that on the VPC level by exposing the list of routing table and find out the main one ? unless in your case the vpc is an external then it will be an issue ?

Reply:
Well, you're right. I can add a check there, and skip this entire thing if `EXTERNAL_RESOURCE_ID` is provided.

Regarding the VPC level: exactly. My original solution was to do this at the VPC level, but that won't work for existing VPC's because there's no callback I can use to populate the `main_route_table_id` runtime property. We spoke about it over the phone...

Reply:
@mabuaisha I added more strict checks.